Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp

```diff
@@ -2466,6 +2468,232 @@ void SpeculativeJIT::emitBranch(Node* node)
+        break;
+    }
+    case UntypedUse: {
+        presentInTable.append(branch64(Equal, entryKeyGPR, keyGPR));
+        // The input key and entryKey are already normalized. So if 64-bit compare fails
+        // and one is not a cell, they're definitely not equal.
+        loopAround.append(branchIfNotCell(entryKeyRegs));
+
+        // entryKey is Cell here.
+        loopAround.append(branchIfNotCell(key.jsValueRegs()));
+        // Both are Cell here.
+        auto bucketIsString = branchIfString(entryKeyGPR);
+        // entryKey is not String here.
+        loopAround.append(branchIfNotHeapBigInt(entryKeyGPR));
+        // entryKey is HeapBigInt here.
+        slowPathCases.append(branchIfHeapBigInt(keyGPR));
+        loopAround.append(jump());
+        // entryKey is String here.
+        bucketIsString.link(this);
+        loopAround.append(branchIfNotString(keyGPR));
+        slowPathCases.append(jump());
+        break;
+    }
+    default:
+        RELEASE_ASSERT_NOT_REACHED();
+    }
+
+    JIT_COMMENT(*this, "The current entryKey doesn't match the target key. Then, get the next entry in the chain and continue.");
+    loopAround.link(this);
+    add32(TrustedImm32(MapOrSet::Accessor::ChainOffset), entryKeyIndexGPR);
+    loadValue(BaseIndex(butterflyOrStorageGPR, entryKeyIndexGPR, TimesEight), entryKeyIndexRegs);
+    jump().linkTo(loop, this);
+
+    JIT_COMMENT(*this, "Found a matched entryKey.");
+    presentInTable.link(this);
+    move(entryKeyIndexGPR, resultGPR);
+    done.append(jump());
+
+    if (!slowPathCases.empty()) {
+        JIT_COMMENT(*this, "The slow path should call the operation.");
+        slowPathCases.link(this);
+        auto operation = std::is_same<MapOrSet, JSMap>::value ? operationMapKeyIndex : operationSetKeyIndex;
+        callOperationWithSilentSpill(operation, resultGPR, LinkableConstant::globalObject(*this, node), mapGPR, keyGPR, hashGPR);
+        exceptionCheck();
+        done.append(jump());
+    }
+
+    JIT_COMMENT(*this, "Didn't find a matched entryKey.");
+    notPresentInTable.link(this);
+    move(TrustedImm32(JSMap::Accessor::InvalidTableIndex), resultGPR);
+
+    done.link(this);
+    strictInt32Result(resultGPR, node);
+}
+
+void SpeculativeJIT::compileMapKeyIndex(Node* node)
+{
+    if (node->child1().useKind() == MapObjectUse)
+        compileGetMapIndexImpl<JSMap>(node);
+    else if (node->child1().useKind() == SetObjectUse)
+        compileGetMapIndexImpl<JSSet>(node);
+    else
+        RELEASE_ASSERT_NOT_REACHED();
+}
+
+void SpeculativeJIT::compileMapValueWithKeyIndex(Node* node)
+{
+    SpeculateCellOperand map(this, node->child1());
+    SpeculateInt32Operand keyIndex(this, node->child2());
+    GPRReg mapGPR = map.gpr();
+    GPRReg keyIndexGPR = keyIndex.gpr();
+
+    GPRTemporary butterflyStorage(this);
+    JSValueRegsTemporary result(this);
+
+    GPRReg butterflyStorageGPR = butterflyStorage.gpr();
+    JSValueRegs resultRegs = result.regs();
```

Comment:
Let's first list all Operands / Temporaries, and calling gpr() etc. later and then call speculate etc.
```
SpeculateCellOperand map(this, node->child1());
SpeculateInt32Operand keyIndex(this, node->child2());
GPRTemporary butterflyStorage(this);
JSValueRegsTemporary result(this);

GPRReg mapGPR = map.gpr();
GPRReg keyIndexGPR = keyIndex.gpr();
GPRReg butterflyStorageGPR = butterflyStorage.gpr();
JSValueRegs resultRegs = result.regs();

speculateMapObject(node->child1(), mapGPR);
```